In [ ]:
!pip install summa

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.9/54.9 kB 1.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for summa: filename=summa-1.2.0-py3-none-any.whl size=54386 sha256=25e070c433e772be786d6d0bf637ced0a3187fdd6bcc266a859f914897f0d35f
  Stored in directory: /root/.cache/pip/wheels/4a/ca/c5/4958614cfba88ed6ceb7cb5a849f9f89f9ac49971616bc919f
Successfully built summa


Ключевые слова

In [ ]:
from summa import keywords
import nltk
from nltk.corpus import stopwords
nltk.download ("stopwords")
stops = list(set(stopwords.words("english")))


text_clean = ""
text =  """The scientific objective of geodesy is to determine the size and shape of
the Earth. The practical role of geodesy is to provide a network of accurately
surveyed points on the Earth’s surface, the vertical elevations and geographic
positions of which are precisely known and, in turn, may be incorporated in
maps. When two geographic coordinates of a control point on the Earth’s
surface, its latitude and longitude, are known as well as its elevation above
sea level, the location of that point is known with accuracy within the limits
of error involved in the surveying processes. In mapping large areas, such as
a whole state or country, the irregularities in the curvature of the Earth must
be considered. A network of precisely surveyed control points provides a
skeleton to which other surveys may be tied to provide progressively finer
networks of more closely space points. The resulting networks of points have
many uses, including anchor points or bench marks for surveys of highways
and other civil features. A major use of control points is to provide reference
points to which the contour lines and other features of topographic maps are
tied. Most topographic maps are made using photogrammetric techniques
and aerial photographs."""
# уберем стоп-слова
for i in text.split():
    if i not in stops:
        text_clean += i + " "
keywords.keywords (text_clean, language = "english").split("\n")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


['surveyed points',
 'surveying',
 'surveys',
 'control point',
 'provides',
 'provide network',
 'networks',
 'maps',
 'mapping',
 'geodesy',
 'precisely known',
 'techniques aerial',
 'features',
 'uses',
 'use']

Ключевые предложения

In [ ]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('stopwords')
from nltk.tokenize import sent_tokenize
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.probability import FreqDist
from nltk.tokenize import PunktSentenceTokenizer
from nltk.corpus import state_union
from nltk.tokenize import PunktSentenceTokenizer
from nltk.corpus import state_union

# Загрузка текста
text = """The scientific objective of geodesy is to determine the size and shape of the Earth. The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps. When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the surveying processes. In mapping large areas, such as a whole state or country, the irregularities in the curvature of the Earth must be considered. A network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points. The resulting networks of points have many uses, including anchor points or bench marks for surveys of highways and other civil features. A major use of control points is to provide reference points to which the contour lines and other features of topographic maps are tied. Most topographic maps are made using photogrammetric techniques and aerial photographs."""



# Токенизация предложений
sentences = sent_tokenize(text)

# Предварительная обработка текста
stop_words = set(stopwords.words("english"))
filtered_sentences = []
for sentence in sentences:
    words = word_tokenize(sentence)
    filtered_sentence = [word for word in words if word.lower() not in stop_words]
    filtered_sentences.append(" ".join(filtered_sentence))

# Вычисление частотности слов
word_frequencies = FreqDist(word_tokenize(" ".join(filtered_sentences)))

# Вычисление значимости предложений с помощью алгоритма TextRank
ranking_sentences = {}
for i, sentence in enumerate(filtered_sentences):
    sentence_tokens = word_tokenize(sentence)
    ranking_sentences[i] = sum([word_frequencies[token] for token in sentence_tokens])

# Сортировка предложений по значимости
sorted_sentences = sorted(ranking_sentences, key=ranking_sentences.get, reverse=True)

# Выбор N наиболее значимых предложений
num_key_sentences = 3
key_sentences = [sentences[index] for index in sorted_sentences[:num_key_sentences]]

for sentence in key_sentences:
    print(sentence)


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps.
When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the surveying processes.
A network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points.


Вопросы вида "заполнить пробел"

In [ ]:
from summa import keywords
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.probability import FreqDist

nltk.download("punkt")
nltk.download("stopwords")

# Загрузка текста
text = """The scientific objective of geodesy is to determine the size and shape of the Earth. The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps. When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the surveying processes. In mapping large areas, such as a whole state or country, the irregularities in the curvature of the Earth must be considered. A network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points. The resulting networks of points have many uses, including anchor points or bench marks for surveys of highways and other civil features. A major use of control points is to provide reference points to which the contour lines and other features of topographic maps are tied. Most topographic maps are made using photogrammetric techniques and aerial photographs."""

# Уберем стоп-слова
stops = set(stopwords.words("english"))
filtered_words = [word for word in word_tokenize(text) if word.lower() not in stops]

# Генерация ключевых слов
key_words = keywords.keywords(" ".join(filtered_words), language="english").split("\n")

# Токенизация предложений
sentences = sent_tokenize(text)

# Предварительная обработка текста
filtered_sentences = []
for sentence in sentences:
    words = word_tokenize(sentence)
    filtered_sentence = [word for word in words if word.lower() not in stops]
    filtered_sentences.append(" ".join(filtered_sentence))

# Вычисление частотности слов
word_frequencies = FreqDist(word_tokenize(" ".join(filtered_sentences)))

# Вычисление значимости предложений с помощью алгоритма TextRank
ranking_sentences = {}
for i, sentence in enumerate(filtered_sentences):
    sentence_tokens = word_tokenize(sentence)
    ranking_sentences[i] = sum([word_frequencies[token] for token in sentence_tokens])

# Сортировка предложений по значимости
sorted_sentences = sorted(ranking_sentences, key=ranking_sentences.get, reverse=True)

# Выбор N наиболее значимых предложений
num_key_sentences = 3
key_sentences = [sentences[index] for index in sorted_sentences[:num_key_sentences]]

# Замена ключевых слов на "_____"
replaced_sentences = []
for sentence in key_sentences:
    sentence_words = word_tokenize(sentence)
    replaced_words = [
        word if word.lower() not in key_words else "_____" for word in sentence_words
    ]
    replaced_sentence = " ".join(replaced_words)
    replaced_sentences.append(replaced_sentence)

# Вывод результирующего текста с замененными ключевыми словами в ключевых предложениях
result_text = " ".join(replaced_sentences)
for sentence in key_sentences:
    result_text = result_text.replace(sentence, sentence.replace("_____", "_____"))

print(result_text)

The practical role of _____ is to provide a network of accurately surveyed points on the _____ ’ s surface , the vertical elevations and geographic positions of which are precisely known and , in turn , may be incorporated in _____ . When two geographic coordinates of a control point on the _____ ’ s surface , its latitude and longitude , are known as well as its elevation above sea level , the location of that point is known with accuracy within the limits of error involved in the _____ processes . A network of precisely surveyed control points _____ a skeleton to which other _____ may be tied to provide progressively finer _____ of more closely spaced points .


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Вопросы вида "сделать одиночный выбор"

In [ ]:
from summa import keywords
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.probability import FreqDist
from nltk.corpus import wordnet

nltk.download("punkt")
nltk.download("stopwords")
nltk.download("wordnet")

# Функция для замены ключевых слов синонимами
def replace_keywords_with_synonyms(keywords, text):
    replaced_text = text
    for keyword in keywords:
        synonyms = []
        synonyms.append(keyword)  # Включаем ключевое слово в список синонимов
        for syn in wordnet.synsets(keyword):
            for lemma in syn.lemmas():
                synonyms.append(lemma.name())
        if synonyms:
            synonyms = list(set(synonyms))  # Удаляем дубликаты синонимов
            replaced_word = " | ".join(synonyms[:4])  # Выбираем первые три синонима из списка
            replaced_text = replaced_text.replace(keyword, replaced_word)
    return replaced_text

# Загрузка текста
text = """The scientific objective of geodesy is to determine the size and shape of the Earth. The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps. When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the surveying processes. In mapping large areas, such as a whole state or country, the irregularities in the curvature of the Earth must be considered. A network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points. The resulting networks of points have many uses, including anchor points or bench marks for surveys of highways and other civil features. A major use of control points is to provide reference points to which the contour lines and other features of topographic maps are tied. Most topographic maps are made using photogrammetric techniques and aerial photographs."""

# Уберем стоп-слова
stops = set(stopwords.words("english"))
filtered_words = [word for word in word_tokenize(text) if word.lower() not in stops]

# Генерация ключевых слов
key_words = keywords.keywords(" ".join(filtered_words), language="english").split("\n")

# Токенизация предложений
sentences = sent_tokenize(text)

# Предварительная обработка текста
filtered_sentences = []
for sentence in sentences:
    words = word_tokenize(sentence)
    filtered_sentence = [word for word in words if word.lower() not in stops]
    filtered_sentences.append(" ".join(filtered_sentence))

# Вычисление частотности слов
word_frequencies = FreqDist(word_tokenize(" ".join(filtered_sentences)))

# Вычисление значимости предложений с помощью алгоритма TextRank
ranking_sentences = {}
for i, sentence in enumerate(filtered_sentences):
    sentence_tokens = word_tokenize(sentence)
    ranking_sentences[i] = sum([word_frequencies[token] for token in sentence_tokens])

# Сортировка предложений по значимости
sorted_sentences = sorted(ranking_sentences, key=ranking_sentences.get, reverse=True)

# Выбор N наиболее значимых предложений
num_key_sentences = 3  # Количество ключевых предложений, которые вы хотите получить
key_sentences = [sentences[index] for index in sorted_sentences[:num_key_sentences]]

# Замена ключевых слов синонимами
replaced_text_with_synonyms = replace_keywords_with_synonyms(key_words, text)

# Вывод результирующего текста с замененными ключевыми словами синонимами
print(replaced_text_with_synonyms)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...


The scientific objective of geodesy is to determine the size and shape of the Earth. The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in function | maps | single-valued_function | function | single-valued_function | chromosome_mapping | mapping. When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the survey | go_over | surveil | follow processes. In function | single-valued_function | chromosome_mapping | mapping large areas, such as a whole state or country, the irregularities in the curvature of the Earth must be considered. A network of precisely surveyed control points provides | provide | allow | leave a skeleton to which ot

In [ ]:
!pip install sentence-transformers transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 171.5/171.5 kB 3.3 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-many

In [ ]:
from sentence_transformers import SentenceTransformer
from transformers import T5ForConditionalGeneration, T5Tokenizer, BertTokenizer, BertModel, AutoTokenizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import torch
import spacy
from transformers import BertTokenizer, BertModel
from warnings import filterwarnings as filt

filt('ignore')

bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained("bert-base-uncased")
model = SentenceTransformer('distilbert-base-nli-mean-tokens')
nlp = spacy.load("en_core_web_sm")

def get_question(sentence, answer):

  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  text = "context: {} answer: {}".format(sentence,answer)
  max_len = 256
  encoding = tknizer.encode_plus(text,max_length=max_len, pad_to_max_length=False,truncation=True, return_tensors="pt")

  input_ids, attention_mask = encoding["input_ids"], encoding["attention_mask"]

  outs = mdl.generate(input_ids=input_ids,
                                  attention_mask=attention_mask,
                                  early_stopping=True,
                                  num_beams=5,
                                  num_return_sequences=1,
                                  no_repeat_ngram_size=2,
                                  max_length=300)


  dec = [tknizer.decode(ids,skip_special_tokens=True) for ids in outs]


  Question = dec[0].replace("question:","")
  Question= Question.strip()
  return Question

def get_embedding(doc):

  bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
  bert_model = BertModel.from_pretrained("bert-base-uncased")

  tokens = bert_tokenizer.tokenize(txt)
  token_idx = bert_tokenizer.convert_tokens_to_ids(tokens)
  segment_ids = [1] * len(tokens)

  torch_token = torch.tensor([token_idx])
  torch_segment = torch.tensor([segment_ids])

  return bert_model(torch_token, torch_segment)[-1].detach().numpy()

def get_pos(context):
  doc = nlp(context)
  docs = [d.pos_ for d in doc]
  return docs, context.split()

def get_sent(context):
  doc = nlp(context)
  return list(doc.sents)

def get_vector(doc):
  stop_words = "english"
  n_gram_range = (1, 1)
  cv = CountVectorizer(ngram_range=n_gram_range, stop_words=stop_words)
  df = cv.fit_transform([doc])
  feature_names = cv.get_feature_names_out()
  return feature_names


def get_key_words(context, module_type = 't'):
  keywords = []
  top_n = 1
  for txt in get_sent(context):
    keywd = get_vector(str(txt))
    print(f'vectors : {keywd}')
    if module_type == 't':
      doc_embedding = get_embedding(str(txt))
      keywd_embedding = get_embedding(' '.join(keywd))
    else:
      doc_embedding = model.encode([str(txt)])
      keywd_embedding = model.encode(keywd)

    distances = cosine_similarity(doc_embedding, keywd_embedding)
    print(distances)
    keywords += [(keywd[index], str(txt)) for index in distances.argsort()[0][-top_n:]]

  return keywords

txt = """The scientific objective of geodesy is to determine the size and shape of
the Earth. The practical role of geodesy is to provide a network of accurately
surveyed points on the Earth’s surface, the vertical elevations and geographic
positions of which are precisely known and, in turn, may be incorporated in
maps. When two geographic coordinates of a control point on the Earth’s
surface, its latitude and longitude, are known as well as its elevation above
sea level, the location of that point is known with accuracy within the limits
of error involved in the surveying processes. In mapping large areas, such as
a whole state or country, the irregularities in the curvature of the Earth must
be considered. A network of precisely surveyed control points provides a
skeleton to which other surveys may be tied to provide progressively finer
networks of more closely space points. The resulting networks of points have
many uses, including anchor points or bench marks for surveys of highways
and other civil features. A major use of control points is to provide reference
points to which the contour lines and other features of topographic maps are
tied. Most topographic maps are made using photogrammetric techniques
and aerial photographs."""
get_key_words(txt)

for ans, context in get_key_words(txt, 'st'):
  print('=======================================')
  print()
  print(get_question(context, ans))
  print()

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/4.02k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/550 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/265M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/450 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

vectors : ['determine' 'earth' 'geodesy' 'objective' 'scientific' 'shape' 'size']
[[0.9999999]]
vectors : ['accurately' 'earth' 'elevations' 'geodesy' 'geographic' 'incorporated'
 'known' 'maps' 'network' 'points' 'positions' 'practical' 'precisely'
 'provide' 'role' 'surface' 'surveyed' 'turn' 'vertical']
[[0.9999999]]
vectors : ['accuracy' 'control' 'coordinates' 'earth' 'elevation' 'error'
 'geographic' 'involved' 'known' 'latitude' 'level' 'limits' 'location'
 'longitude' 'point' 'processes' 'sea' 'surface' 'surveying']
[[0.9999999]]
vectors : ['areas' 'considered' 'country' 'curvature' 'earth' 'irregularities'
 'large' 'mapping' 'state']
[[0.9999999]]
vectors : ['closely' 'control' 'finer' 'network' 'networks' 'points' 'precisely'
 'progressively' 'provide' 'provides' 'skeleton' 'space' 'surveyed'
 'surveys' 'tied']
[[0.9999999]]
vectors : ['anchor' 'bench' 'civil' 'features' 'highways' 'including' 'marks'
 'networks' 'points' 'resulting' 'surveys' 'uses']
[[0.9999999]]
vectors : 

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/892M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/1.86k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.79k [00:00<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


What is the scientific objective of determining the size and shape of the Earth?


What provides a network of accurately surveyed points on the Earth's surface?


The location of a control point is known with what within the limits of error?


What is it called when the curvature of the Earth is considered?


A network of precisely what type of control points provides a skeleton to which other surveys may be tied?


Anchor points can be used for surveys of what?


Control points are used to tie contour lines and other features of what?


What type of techniques are most topographic maps made using?



In [ ]:
q = []
a = []
for ans, context in get_key_words(txt, 'st'):
  print('=======================================')
  print()
  q.append(get_question(context, ans))
  a.append(ans)
  print(get_question(context, ans))
  print()
  print(ans)

vectors : ['determine' 'earth' 'geodesy' 'objective' 'scientific' 'shape' 'size']
[[0.31028125 0.47714534 0.7102637  0.2873761  0.442575   0.36717612
  0.37526548]]
vectors : ['accurately' 'earth' 'elevations' 'geodesy' 'geographic' 'incorporated'
 'known' 'maps' 'network' 'points' 'positions' 'practical' 'precisely'
 'provide' 'role' 'surface' 'surveyed' 'turn' 'vertical']
[[0.37110436 0.34753013 0.39051688 0.6635004  0.44378665 0.23944339
  0.26932943 0.5231131  0.2732569  0.25317523 0.27121705 0.42150202
  0.30519193 0.25051895 0.25342935 0.33382377 0.4248231  0.17559204
  0.302558  ]]
vectors : ['accuracy' 'control' 'coordinates' 'earth' 'elevation' 'error'
 'geographic' 'involved' 'known' 'latitude' 'level' 'limits' 'location'
 'longitude' 'point' 'processes' 'sea' 'surface' 'surveying']
[[0.5339749  0.28389823 0.4015081  0.3564523  0.49106193 0.25222066
  0.45013607 0.25824064 0.3253577  0.5236009  0.3313922  0.25892657
  0.32454252 0.49544552 0.30758154 0.26432085 0.30551773 0.3

In [ ]:
questions = q
answers = a
print(answers)
print(questions)

['geodesy', 'geodesy', 'accuracy', 'mapping', 'surveyed', 'highways', 'maps', 'photogrammetric']
['What is the scientific objective of determining the size and shape of the Earth?', "What provides a network of accurately surveyed points on the Earth's surface?", 'The location of a control point is known with what within the limits of error?', 'What is it called when the curvature of the Earth is considered?', 'A network of precisely what type of control points provides a skeleton to which other surveys may be tied?', 'Anchor points can be used for surveys of what?', 'Control points are used to tie contour lines and other features of what?', 'What type of techniques are most topographic maps made using?']


In [ ]:
questions = q
answers = a
print(answers)
print(questions)

from transformers import T5ForConditionalGeneration, T5Tokenizer

# Загрузка предобученной модели T5
model = T5ForConditionalGeneration.from_pretrained('t5-base')
tokenizer = T5Tokenizer.from_pretrained('t5-base')

# Функция для генерации дистракторов
def generate_distractors(question, context, num_distractors=3):
    input_text = 'question: {} context: {}'.format(question, context)

    # Генерация дистракторов
    inputs = tokenizer.encode(input_text, return_tensors='pt')
    outputs = model.generate(inputs, num_return_sequences=num_distractors, max_length=100, num_beams=5)
    distractors = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]

    return distractors

# Пример использования
for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nВопрос:", question)
    print("Правильный ответ:", answer)

    distractors = generate_distractors(question, text)
    print("Дистракторы:")
    for distractor in distractors:
        print("- " + distractor)

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.



Вопрос: What is the scientific objective of determining the size and shape of the Earth?
Правильный ответ: geodesy
Дистракторы:
- geodesy
- to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps
- to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and may be incorporated in maps

Вопрос: What provides a network of accurately surveyed points on the Earth's surface?
Правильный ответ: geodesy
Дистракторы:
- geodesy
- a network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points
- a skeleton to which other surveys may be tied to provide progressively finer networks of more closely spaced points

Вопрос: The location of a control point is known

In [ ]:
from transformers import T5ForConditionalGeneration, T5Tokenizer
from nltk.corpus import wordnet
import random

# Загрузка предобученной модели T5
model = T5ForConditionalGeneration.from_pretrained('t5-base')
tokenizer = T5Tokenizer.from_pretrained('t5-base')

# Функция для получения синонимов слова
def get_synonyms(word):
    synonyms = set()
    for synset in wordnet.synsets(word):
        for lemma in synset.lemmas():
            synonyms.add(lemma.name())
    return list(synonyms)

# Функция для генерации дистракторов с использованием аугментации данных
def generate_distractors(question, context, num_distractors=3):
    input_text = 'question: {} context: {}'.format(question, context)

    # Генерация дистракторов
    inputs = tokenizer.encode(input_text, return_tensors='pt')
    outputs = model.generate(inputs, num_return_sequences=num_distractors, max_length=100, num_beams=5)
    distractors = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]

    # Добавление синонимов
    augmented_distractors = []
    for distractor in distractors:
        tokens = distractor.split()
        augmented_tokens = []
        for token in tokens:
            synonyms = get_synonyms(token)
            if synonyms:
                augmented_tokens.append(random.choice(synonyms))
            else:
                augmented_tokens.append(token)
        augmented_distractor = ' '.join(augmented_tokens)
        augmented_distractors.append(augmented_distractor)

    return augmented_distractors

# Пример использования
for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nВопрос:", question)
    print("Правильный ответ:", answer)

    distractors = generate_distractors(question, text)
    print("Дистракторы:")
    for distractor in distractors:
        print("- " + distractor)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.



Вопрос: What is the scientific objective of determining the size and shape of the Earth?
Правильный ответ: geodesy
Дистракторы:
- geodesy
- to allow_for type_A mesh of accurately appraise manoeuver on the Earth’s surface, the upright superlative and geographical pose of which exist exactly get_it_on and, IN turn, may constitute incorporate In mapping
- to put_up group_A electronic_network of accurately follow period along the Earth’s surface, the vertical acme and geographic spatial_relation of which make_up on_the_button have_a_go_at_it and whitethorn constitute unified indium map_out

Вопрос: What provides a network of accurately surveyed points on the Earth's surface?
Правильный ответ: geodesy
Дистракторы:
- geodesy
- angstrom_unit net of exactly follow ascendency indicate furnish deoxyadenosine_monophosphate skeleton_in_the_closet to which other follow English_hawthorn embody tied to furnish more_and_more fine electronic_network of to_a_greater_extent close separated spot
- axerop

In [ ]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [ ]:
from nltk.corpus import wordnet

# Function to generate synonyms of the answer
def generate_synonyms(answer, num_synonyms=3):
    synonyms = []
    for syn in wordnet.synsets(answer):
        for lemma in syn.lemmas():
            synonym = lemma.name()
            if synonym != answer:
                synonyms.append(synonym)
                if len(synonyms) == num_synonyms:
                    return synonyms
    return synonyms

for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nQuestion:", question)
    print("Correct answer:", answer)

    synonyms = generate_synonyms(answer)
    print("Synonyms:")
    for synonym in synonyms:
        print("- " + synonym)


Question: What is the scientific objective of determining the size and shape of the Earth?
Correct answer: geodesy
Synonyms:

Question: What provides a network of accurately surveyed points on the Earth's surface?
Correct answer: geodesy
Synonyms:

Question: The location of a control point is known with what within the limits of error?
Correct answer: accuracy
Synonyms:
- truth

Question: What is it called when the curvature of the Earth is considered?
Correct answer: mapping
Synonyms:
- function
- mathematical_function
- single-valued_function

Question: A network of precisely what type of control points provides a skeleton to which other surveys may be tied?
Correct answer: surveyed
Synonyms:
- survey
- appraise
- survey

Question: Anchor points can be used for surveys of what?
Correct answer: highways
Synonyms:
- highway
- main_road

Question: Control points are used to tie contour lines and other features of what?
Correct answer: maps
Synonyms:
- map
- function
- mathematical_func

Вопросы с одиночным выбором

Вопросы с множественным выбором

In [ ]:
pip install transformers==4.8.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 11.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 897.5/897.5 kB 18.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 212.7/212.7 kB 19.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  error: subprocess-exited-with-error
  
  × Building wheel for tokenizers (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for tokenizers
Failed to build tokenizers
ERROR: Could not build wheels for tokenizers, which is required to install pyproject.toml-based projects


In [ ]:
from transformers import T5Tokenizer, T5ForConditionalGeneration

tokenizer = T5Tokenizer.from_pretrained('t5-base')
model = T5ForConditionalGeneration.from_pretrained('t5-base')

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [ ]:
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration

def generate_multiple_choice_question(sentence, correct_answers, num_distractors=3):
    tokenizer = T5Tokenizer.from_pretrained('t5-base')
    model = T5ForConditionalGeneration.from_pretrained('t5-base')

    prompt = f"Question: {sentence}\nOptions:\n"

    # Add correct answers
    for i, ans in enumerate(correct_answers, start=1):
        prompt += f"{i}. {ans}\n"

    # Generate distractors
    with torch.no_grad():
        input_text = prompt + "Distractors:"
        input_ids = tokenizer.encode(input_text, return_tensors='pt')
        output = model.generate(input_ids, max_length=50, num_return_sequences=1)[0]

    generated_text = tokenizer.decode(output, skip_special_tokens=True).split(":")[-1]
    distractors = generated_text.strip().split("\n")

    # Take the first num_distractors distractors
    distractors = distractors[:num_distractors]

    return distractors

sentence = "The scientific objective of geodesy is to determine the size and shape of the Earth."
correct_answers = ["Size of the Earth", "Shape of the Earth"]
distractors = generate_multiple_choice_question(sentence, correct_answers, num_distractors=3)

print("Question:", sentence)
print("Options:")
for i, ans in enumerate(correct_answers, start=1):
    print(f"{i}. {ans}")
for i, distractor in enumerate(distractors, start=len(correct_answers)+1):
    print(f"{i}. {distractor}")

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Question: The scientific objective of geodesy is to determine the size and shape of the Earth.
Options:
1. Size of the Earth
2. Shape of the Earth
3. True


Вопросы с одиночным выбором (да/нет)


In [ ]:
def get_question_with_binary_choice(sentence, answer):
  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  text = "context: {} answer: {}".format(sentence, answer)
  max_len = 256
  encoding = tknizer.encode_plus(text, max_length=max_len, pad_to_max_length=False, truncation=True, return_tensors="pt")

  input_ids, attention_mask = encoding["input_ids"], encoding["attention_mask"]

  outs = mdl.generate(input_ids=input_ids,
                      attention_mask=attention_mask,
                      early_stopping=True,
                      num_beams=5,
                      num_return_sequences=1,
                      no_repeat_ngram_size=2,
                      max_length=300)

  dec = [tknizer.decode(ids, skip_special_tokens=True) for ids in outs]

  question = dec[0].replace("question:", "")
  question = question.strip()


  return question

# Пример использования функции get_question_with_binary_choice
question = get_question_with_binary_choice("The scientific objective of geodesy is to determine the size and shape of the Earth.", "No")
print(question)

Does geodesy determine the size and shape of the Earth?


In [ ]:
def get_question_with_binary_choice(sentence, answer):
  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  text = "context: {} answer: {}".format(sentence, answer)
  max_len = 256
  encoding = tknizer.encode_plus(text, max_length=max_len, pad_to_max_length=False, truncation=True, return_tensors="pt")

  input_ids, attention_mask = encoding["input_ids"], encoding["attention_mask"]

  outs = mdl.generate(input_ids=input_ids,
                      attention_mask=attention_mask,
                      early_stopping=True,
                      num_beams=5,
                      num_return_sequences=1,
                      no_repeat_ngram_size=2,
                      max_length=300)

  dec = [tknizer.decode(ids, skip_special_tokens=True) for ids in outs]

  question = dec[0].replace("question:", "")
  question = question.strip()


  return question

# Пример использования функции get_question_with_binary_choice
question = get_question_with_binary_choice("The scientific objective of geodesy is to determine the size and shape of the Earth.", "No")
print(question)

def generate_binary_choice_questions(text, num_questions):
  questions = []
  sentences = get_sent(text)  # Получение предложений из текста

  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  for sentence in sentences:
    question = get_question_with_binary_choice(sentence, "Yes")  # Генерация вопроса с одиночным выбором
    questions.append(question)

    if len(questions) >= num_questions:
      break

  return questions[:num_questions]


# Пример использования функции generate_binary_choice_questions
text = "The scientific objective of geodesy is to determine the size and shape of the Earth. The practical role of geodesy is to provide a network of accurately surveyed points on the Earth’s surface, the vertical elevations and geographic positions of which are precisely known and, in turn, may be incorporated in maps. When two geographic coordinates of a control point on the Earth’s surface, its latitude and longitude, are known as well as its elevation above sea level, the location of that point is known with accuracy within the limits of error involved in the surveying processes. In mapping large areas, such as a whole state or country, the irregularities in the curvature of the Earth must be considered. A network of precisely surveyed control points provides a skeleton to which other surveys may be tied to provide progressively finer networks of more closely space points. The resulting networks of points have many uses, including anchor points or bench marks for surveys of highways and other civil features. A major use of control points is to provide reference points to which the contour lines and other features of topographic maps are tied. Most topographic maps are made using photogrammetric techniques and aerial photographs."

num_questions = 5
binary_choice_questions = generate_binary_choice_questions(text, num_questions)

for question in binary_choice_questions:
  print(question)

Does geodesy determine the size and shape of the Earth?
Can geodesy be used to create maps?
Is the location of a control point known with accuracy within the limits of error involved in surveying processes?
Do irregularities in the curvature of the Earth must be considered when mapping large areas?
Does a network of precisely surveyed control points provide skeleton to which other surveys may be tied?


ABOUT 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being in what is known as the Big Bang. The story of these fundamental features of our universe is called physics.
About 300,000 years after their appearance, matter and energy started to coalesce into complex structures, called atoms, which then combined into molecules. The story of atoms, molecules and their interactions is called chemistry.
About 3.8 billion years ago, on a planet called Earth, certain molecules combined to form particularly large and intricate structures called organisms. The story of organisms is called biology.
About 70,000 years ago, organisms belonging to the species Homo sapiens started to form even more elaborate structures called cultures. The subsequent development of these human cultures is called history.
Three important revolutions shaped the course of history: the Cognitive Revolution kick-started history about 70,000 years ago. The Agricultural Revolution sped it up about 12,000 years ago. The Scientific Revolution, which got under way only 500 years ago, may well end history and start something completely different. This book tells the story of how these three revolutions have affected humans and their fellow organisms.


In [ ]:
from summa import keywords
import nltk
from nltk.corpus import stopwords
nltk.download ("stopwords")
stops = list(set(stopwords.words("english")))


text_clean = ""
text =  """ABOUT 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being in what is known as the Big Bang. The story of these fundamental features of our
universe is called physics.
About 300,000 years after their appearance, matter and energy started to coalesce into complex structures, called atoms, which then combined into molecules.
The story of atoms, molecules and their interactions is called chemistry.
About 3.8 billion years ago, on a planet called Earth, certain molecules combined to form particularly large and intricate structures called organisms.
The story of organisms is called biology.
About 70,000 years ago, organisms belonging to the species Homo sapiens started to form even more elaborate structures called cultures.
The subsequent development of these human cultures is called history.
Three important revolutions shaped the course of history: the Cognitive Revolution kick-started history about 70,000 years ago.
The Agricultural Revolution sped it up about 12,000 years ago. The Scientific Revolution,
which got under way only 500 years ago, may well end history and start something completely different.
This book tells the story of how these three revolutions have
affected humans and their fellow organisms.."""
# уберем стоп-слова
for i in text.split():
    if i not in stops:
        text_clean += i + " "
keywords.keywords (text_clean, language = "english").split("\n")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['called',
 'revolutions',
 'revolution',
 'years',
 'started',
 'start',
 'history',
 'organisms',
 'structures',
 'human',
 'humans',
 'story',
 'molecules',
 'ago',
 'atoms']

In [ ]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('stopwords')
from nltk.tokenize import sent_tokenize
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.probability import FreqDist
from nltk.tokenize import PunktSentenceTokenizer
from nltk.corpus import state_union
from nltk.tokenize import PunktSentenceTokenizer
from nltk.corpus import state_union

# Загрузка текста
text =  """ABOUT 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being in what is known as the Big Bang. The story of these fundamental features of our
universe is called physics.
About 300,000 years after their appearance, matter and energy started to coalesce into complex structures, called atoms, which then combined into molecules.
The story of atoms, molecules and their interactions is called chemistry.
About 3.8 billion years ago, on a planet called Earth, certain molecules combined to form particularly large and intricate structures called organisms.
The story of organisms is called biology.
About 70,000 years ago, organisms belonging to the species Homo sapiens started to form even more elaborate structures called cultures.
The subsequent development of these human cultures is called history.
Three important revolutions shaped the course of history: the Cognitive Revolution kick-started history about 70,000 years ago.
The Agricultural Revolution sped it up about 12,000 years ago. The Scientific Revolution,
which got under way only 500 years ago, may well end history and start something completely different.
This book tells the story of how these three revolutions have
affected humans and their fellow organisms.."""


# Токенизация предложений
sentences = sent_tokenize(text)

# Предварительная обработка текста
stop_words = set(stopwords.words("english"))
filtered_sentences = []
for sentence in sentences:
    words = word_tokenize(sentence)
    filtered_sentence = [word for word in words if word.lower() not in stop_words]
    filtered_sentences.append(" ".join(filtered_sentence))

# Вычисление частотности слов
word_frequencies = FreqDist(word_tokenize(" ".join(filtered_sentences)))

# Вычисление значимости предложений с помощью алгоритма TextRank
ranking_sentences = {}
for i, sentence in enumerate(filtered_sentences):
    sentence_tokens = word_tokenize(sentence)
    ranking_sentences[i] = sum([word_frequencies[token] for token in sentence_tokens])

# Сортировка предложений по значимости
sorted_sentences = sorted(ranking_sentences, key=ranking_sentences.get, reverse=True)

# Выбор N наиболее значимых предложений
num_key_sentences = 3
key_sentences = [sentences[index] for index in sorted_sentences[:num_key_sentences]]

for sentence in key_sentences:
    print(sentence)


About 3.8 billion years ago, on a planet called Earth, certain molecules combined to form particularly large and intricate structures called organisms.
About 300,000 years after their appearance, matter and energy started to coalesce into complex structures, called atoms, which then combined into molecules.
The Scientific Revolution, 
which got under way only 500 years ago, may well end history and start something completely different.


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Вопросы вида заполнить пробел

In [ ]:
from summa import keywords
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.probability import FreqDist

nltk.download("punkt")
nltk.download("stopwords")

# Уберем стоп-слова
stops = set(stopwords.words("english"))
filtered_words = [word for word in word_tokenize(text) if word.lower() not in stops]

# Генерация ключевых слов
key_words = keywords.keywords(" ".join(filtered_words), language="english").split("\n")

# Токенизация предложений
sentences = sent_tokenize(text)

# Предварительная обработка текста
filtered_sentences = []
for sentence in sentences:
    words = word_tokenize(sentence)
    filtered_sentence = [word for word in words if word.lower() not in stops]
    filtered_sentences.append(" ".join(filtered_sentence))

# Вычисление частотности слов
word_frequencies = FreqDist(word_tokenize(" ".join(filtered_sentences)))

# Вычисление значимости предложений с помощью алгоритма TextRank
ranking_sentences = {}
for i, sentence in enumerate(filtered_sentences):
    sentence_tokens = word_tokenize(sentence)
    ranking_sentences[i] = sum([word_frequencies[token] for token in sentence_tokens])

# Сортировка предложений по значимости
sorted_sentences = sorted(ranking_sentences, key=ranking_sentences.get, reverse=True)

# Выбор N наиболее значимых предложений
num_key_sentences = 3
key_sentences = [sentences[index] for index in sorted_sentences[:num_key_sentences]]

# Замена ключевых слов на "_____"
replaced_sentences = []
for sentence in key_sentences:
    sentence_words = word_tokenize(sentence)
    replaced_words = [
        word if word.lower() not in key_words else "_____" for word in sentence_words
    ]
    replaced_sentence = " ".join(replaced_words)
    replaced_sentences.append(replaced_sentence)

# Вывод результирующего текста с замененными ключевыми словами в ключевых предложениях
result_text = " ".join(replaced_sentences)
for sentence in key_sentences:
    result_text = result_text.replace(sentence, sentence.replace("_____", "_____"))

print(result_text)

About 3.8 billion _____ _____ , on a planet _____ Earth , certain _____ combined to form particularly large and intricate _____ _____ _____ . About 300,000 _____ after their appearance , matter and energy _____ to coalesce into complex _____ , _____ _____ , which then combined into _____ . The Scientific _____ , which got under way only 500 _____ _____ , may well end _____ and _____ something completely different .


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Открытые вопросы

In [ ]:
from sentence_transformers import SentenceTransformer
from transformers import T5ForConditionalGeneration, T5Tokenizer, BertTokenizer, BertModel, AutoTokenizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import torch
import spacy
from transformers import BertTokenizer, BertModel
from warnings import filterwarnings as filt

filt('ignore')

bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = BertModel.from_pretrained("bert-base-uncased")
model = SentenceTransformer('distilbert-base-nli-mean-tokens')
nlp = spacy.load("en_core_web_sm")

def get_question(sentence, answer):

  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  text = "context: {} answer: {}".format(sentence,answer)
  max_len = 256
  encoding = tknizer.encode_plus(text,max_length=max_len, pad_to_max_length=False,truncation=True, return_tensors="pt")

  input_ids, attention_mask = encoding["input_ids"], encoding["attention_mask"]

  outs = mdl.generate(input_ids=input_ids,
                                  attention_mask=attention_mask,
                                  early_stopping=True,
                                  num_beams=5,
                                  num_return_sequences=1,
                                  no_repeat_ngram_size=2,
                                  max_length=300)


  dec = [tknizer.decode(ids,skip_special_tokens=True) for ids in outs]


  Question = dec[0].replace("question:","")
  Question= Question.strip()
  return Question

def get_embedding(doc):

  bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
  bert_model = BertModel.from_pretrained("bert-base-uncased")

  tokens = bert_tokenizer.tokenize(txt)
  token_idx = bert_tokenizer.convert_tokens_to_ids(tokens)
  segment_ids = [1] * len(tokens)

  torch_token = torch.tensor([token_idx])
  torch_segment = torch.tensor([segment_ids])

  return bert_model(torch_token, torch_segment)[-1].detach().numpy()

def get_pos(context):
  doc = nlp(context)
  docs = [d.pos_ for d in doc]
  return docs, context.split()

def get_sent(context):
  doc = nlp(context)
  return list(doc.sents)

def get_vector(doc):
  stop_words = "english"
  n_gram_range = (1, 1)
  cv = CountVectorizer(ngram_range=n_gram_range, stop_words=stop_words)
  df = cv.fit_transform([doc])
  feature_names = cv.get_feature_names_out()
  return feature_names


def get_key_words(context, module_type = 't'):
  keywords = []
  top_n = 1
  for txt in get_sent(context):
    keywd = get_vector(str(txt))
    print(f'vectors : {keywd}')
    if module_type == 't':
      doc_embedding = get_embedding(str(txt))
      keywd_embedding = get_embedding(' '.join(keywd))
    else:
      doc_embedding = model.encode([str(txt)])
      keywd_embedding = model.encode(keywd)

    distances = cosine_similarity(doc_embedding, keywd_embedding)
    print(distances)
    keywords += [(keywd[index], str(txt)) for index in distances.argsort()[0][-top_n:]]

  return keywords

txt = text
get_key_words(txt)

for ans, context in get_key_words(txt, 'st'):
  print('=======================================')
  print()
  print(get_question(context, ans))
  print()

vectors : ['13' 'ago' 'bang' 'big' 'billion' 'came' 'energy' 'known' 'matter'
 'space' 'time' 'years']
[[1.]]
vectors : ['called' 'features' 'fundamental' 'physics' 'story' 'universe']
[[1.]]
vectors : ['000' '300' 'appearance' 'atoms' 'called' 'coalesce' 'combined' 'complex'
 'energy' 'matter' 'molecules' 'started' 'structures' 'years']
[[1.]]
vectors : ['atoms' 'called' 'chemistry' 'interactions' 'molecules' 'story']
[[1.]]
vectors : ['ago' 'billion' 'called' 'certain' 'combined' 'earth' 'form' 'intricate'
 'large' 'molecules' 'organisms' 'particularly' 'planet' 'structures'
 'years']
[[1.]]
vectors : ['biology' 'called' 'organisms' 'story']
[[1.]]
vectors : ['000' '70' 'ago' 'belonging' 'called' 'cultures' 'elaborate' 'form'
 'homo' 'organisms' 'sapiens' 'species' 'started' 'structures' 'years']
[[1.]]
vectors : ['called' 'cultures' 'development' 'history' 'human' 'subsequent']
[[1.]]
vectors : ['000' '70' 'ago' 'cognitive' 'course' 'history' 'important' 'kick'
 'revolution' 'revolu

In [ ]:
q = []
a = []
for ans, context in get_key_words(txt, 'st'):
  print('=======================================')
  print()
  q.append(get_question(context, ans))
  a.append(ans)
  print(get_question(context, ans))
  print()
  print(ans)

vectors : ['13' 'ago' 'bang' 'big' 'billion' 'came' 'energy' 'known' 'matter'
 'space' 'time' 'years']
[[0.40240687 0.32132065 0.4152053  0.44846058 0.4285241  0.34803778
  0.35654643 0.22459406 0.30112398 0.3977958  0.29370135 0.42446303]]
vectors : ['called' 'features' 'fundamental' 'physics' 'story' 'universe']
[[0.32055312 0.38015568 0.45640394 0.8372833  0.38811198 0.5041977 ]]
vectors : ['000' '300' 'appearance' 'atoms' 'called' 'coalesce' 'combined' 'complex'
 'energy' 'matter' 'molecules' 'started' 'structures' 'years']
[[0.27816117 0.4709214  0.32143778 0.445777   0.21099615 0.4300646
  0.33059615 0.30794853 0.35984766 0.22374323 0.42215937 0.307
  0.31950516 0.3957219 ]]
vectors : ['atoms' 'called' 'chemistry' 'interactions' 'molecules' 'story']
[[0.6301067  0.3697598  0.8131239  0.5206313  0.67972916 0.42763355]]
vectors : ['ago' 'billion' 'called' 'certain' 'combined' 'earth' 'form' 'intricate'
 'large' 'molecules' 'organisms' 'particularly' 'planet' 'structures'
 'years']


Вопросы с множественным выбором

In [ ]:
questions = q
answers = a
print(answers)
print(questions)

from transformers import T5ForConditionalGeneration, T5Tokenizer

# Загрузка предобученной модели T5
model = T5ForConditionalGeneration.from_pretrained('t5-base')
tokenizer = T5Tokenizer.from_pretrained('t5-base')

# Функция для генерации дистракторов
def generate_distractors(question, context, num_distractors=3):
    input_text = 'question: {} context: {}'.format(question, context)

    # Генерация дистракторов
    inputs = tokenizer.encode(input_text, return_tensors='pt')
    outputs = model.generate(inputs, num_return_sequences=num_distractors, max_length=100, num_beams=5)
    distractors = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]

    return distractors

# Пример использования
for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nВопрос:", question)
    print("Правильный ответ:", answer)

    distractors = generate_distractors(question, text)
    print("Дистракторы:")
    for distractor in distractors:
        print("- " + distractor)

['big', 'physics', '300', 'chemistry', 'planet', 'biology', '70', 'history', 'years', 'agricultural', 'end', 'revolutions']
['How big is the Big Bang?', 'What is the story of fundamental features of our universe called?', 'About how many years after their appearance did matter and energy begin to coalesce into complex structures?', 'What is the story of atoms, molecules and interactions called?', 'On what planet did organisms form?', 'What is the name of the story of organisms?', 'About how many years ago did organisms belonging to Homo sapiens start to form cultures?', 'What is the development of human cultures called?', 'How long ago did the Cognitive Revolution last?', 'What revolution sped up agriculture 12,000 years ago?', 'What could happen to history with the Scientific Revolution?', 'What has affected humans and their fellow organisms?']


Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.



Вопрос: How big is the Big Bang?
Правильный ответ: big
Дистракторы:
- 13.5 BILLION YEARS AGO
- 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being
- 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being in what is known as the Big Bang

Вопрос: What is the story of fundamental features of our universe called?
Правильный ответ: physics
Дистракторы:
- physics
- physics.
- physics.

Вопрос: About how many years after their appearance did matter and energy begin to coalesce into complex structures?
Правильный ответ: 300
Дистракторы:
- 300,000 years
- 300,000
- 300,000 years after their appearance

Вопрос: What is the story of atoms, molecules and interactions called?
Правильный ответ: chemistry
Дистракторы:
- chemistry
- chemistry.
- chemistry.

Вопрос: On what planet did organisms form?
Правильный ответ: planet
Дистракторы:
- Earth
- Earth 3.8 billion years ago
- Earth 3.8 billion years ago,

Вопрос: What is the name of the story of organisms?
Правильн

In [ ]:
from transformers import T5ForConditionalGeneration, T5Tokenizer
from nltk.corpus import wordnet
import random

# Загрузка предобученной модели T5
model = T5ForConditionalGeneration.from_pretrained('t5-base')
tokenizer = T5Tokenizer.from_pretrained('t5-base')

# Функция для получения синонимов слова
def get_synonyms(word):
    synonyms = set()
    for synset in wordnet.synsets(word):
        for lemma in synset.lemmas():
            synonyms.add(lemma.name())
    return list(synonyms)

# Функция для генерации дистракторов с использованием аугментации данных
def generate_distractors(question, context, num_distractors=3):
    input_text = 'question: {} context: {}'.format(question, context)

    # Генерация дистракторов
    inputs = tokenizer.encode(input_text, return_tensors='pt')
    outputs = model.generate(inputs, num_return_sequences=num_distractors, max_length=100, num_beams=5)
    distractors = [tokenizer.decode(output, skip_special_tokens=True) for output in outputs]

    # Добавление синонимов
    augmented_distractors = []
    for distractor in distractors:
        tokens = distractor.split()
        augmented_tokens = []
        for token in tokens:
            synonyms = get_synonyms(token)
            if synonyms:
                augmented_tokens.append(random.choice(synonyms))
            else:
                augmented_tokens.append(token)
        augmented_distractor = ' '.join(augmented_tokens)
        augmented_distractors.append(augmented_distractor)

    return augmented_distractors

# Пример использования
for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nВопрос:", question)
    print("Правильный ответ:", answer)

    distractors = generate_distractors(question, text)
    print("Дистракторы:")
    for distractor in distractors:
        print("- " + distractor)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.



Вопрос: How big is the Big Bang?
Правильный ответ: big
Дистракторы:
- 13.5 million year agone
- 13.5 jillion twelvemonth AGO, MATTER, energy, meter and blank_space total into equal
- 13.5 one_thousand_million yr AGO, MATTER, energy, time and infinite occur into personify inch what follow roll_in_the_hay antiophthalmic_factor the self-aggrandizing fringe

Вопрос: What is the story of fundamental features of our universe called?
Правильный ответ: physics
Дистракторы:
- physical_science
- physics.
- physics.

Вопрос: About how many years after their appearance did matter and energy begin to coalesce into complex structures?
Правильный ответ: 300
Дистракторы:
- 300,000 geezerhood
- 300,000
- 300,000 twelvemonth afterwards their appearing

Вопрос: What is the story of atoms, molecules and interactions called?
Правильный ответ: chemistry
Дистракторы:
- alchemy
- chemistry.
- chemistry.

Вопрос: On what planet did organisms form?
Правильный ответ: planet
Дистракторы:
- solid_ground
- worldly

In [ ]:
from nltk.corpus import wordnet

# Function to generate synonyms of the answer
def generate_synonyms(answer, num_synonyms=3):
    synonyms = []
    for syn in wordnet.synsets(answer):
        for lemma in syn.lemmas():
            synonym = lemma.name()
            if synonym != answer:
                synonyms.append(synonym)
                if len(synonyms) == num_synonyms:
                    return synonyms
    return synonyms

for i in range(len(questions)):
    question = questions[i]
    answer = answers[i]

    print("\nQuestion:", question)
    print("Correct answer:", answer)

    synonyms = generate_synonyms(answer)
    print("Synonyms:")
    for synonym in synonyms:
        print("- " + synonym)


Question: How big is the Big Bang?
Correct answer: big
Synonyms:
- large
- bad
- large

Question: What is the story of fundamental features of our universe called?
Correct answer: physics
Synonyms:
- natural_philosophy
- physical_science
- purgative

Question: About how many years after their appearance did matter and energy begin to coalesce into complex structures?
Correct answer: 300
Synonyms:
- three_hundred
- ccc

Question: What is the story of atoms, molecules and interactions called?
Correct answer: chemistry
Synonyms:
- chemical_science
- interpersonal_chemistry
- alchemy

Question: On what planet did organisms form?
Correct answer: planet
Synonyms:
- major_planet
- satellite

Question: What is the name of the story of organisms?
Correct answer: biology
Synonyms:
- biological_science
- biota

Question: About how many years ago did organisms belonging to Homo sapiens start to form cultures?
Correct answer: 70
Synonyms:
- seventy
- LXX
- seventy

Question: What is the developmen

In [ ]:
def get_question_with_binary_choice(sentence, answer):
  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  text = "context: {} answer: {}".format(sentence, answer)
  max_len = 256
  encoding = tknizer.encode_plus(text, max_length=max_len, pad_to_max_length=False, truncation=True, return_tensors="pt")

  input_ids, attention_mask = encoding["input_ids"], encoding["attention_mask"]

  outs = mdl.generate(input_ids=input_ids,
                      attention_mask=attention_mask,
                      early_stopping=True,
                      num_beams=5,
                      num_return_sequences=1,
                      no_repeat_ngram_size=2,
                      max_length=300)

  dec = [tknizer.decode(ids, skip_special_tokens=True) for ids in outs]

  question = dec[0].replace("question:", "")
  question = question.strip()


  return question


def generate_binary_choice_questions(text, num_questions):
  questions = []
  sentences = get_sent(text)  # Получение предложений из текста

  mdl = T5ForConditionalGeneration.from_pretrained('ramsrigouthamg/t5_squad_v1')
  tknizer = AutoTokenizer.from_pretrained('ramsrigouthamg/t5_squad_v1')

  for sentence in sentences:
    question = get_question_with_binary_choice(sentence, "Yes")  # Генерация вопроса с одиночным выбором
    questions.append(question)

    if len(questions) >= num_questions:
      break

  return questions[:num_questions]


num_questions = 5
binary_choice_questions = generate_binary_choice_questions(text, num_questions)

for question in binary_choice_questions:
  print(question)

Did matter, energy, time and space come into being in the Big Bang?
Does physics tell the story of fundamental features of our universe?
Did matter and energy start to coalesce into complex structures?
Does chemistry tell the story of atoms, molecules and interactions?
Did organisms form on Earth?


ABOUT 13.5 BILLION YEARS AGO, MATTER, energy, time and space came into being in what is known as the Big Bang. The story of these fundamental features of our universe is called physics.

About 300,000 years after their appearance, matter and energy started to coalesce into complex structures, called atoms, which then combined into molecules. The story of atoms, molecules and their interactions is called chemistry.

About 3.8 billion years ago, on a planet called Earth, certain molecules combined to form particularly large and intricate structures called organisms. The story of organisms is called biology.

About 70,000 years ago, organisms belonging to the species Homo sapiens started to form even more elaborate structures called cultures. The subsequent development of these human cultures is called history.

Three important revolutions shaped the course of history: the Cognitive Revolution kick-started history about 70,000 years ago. The Agricultural Revolution sped it up about 12,000 years ago. The Scientific Revolution, which got under way only 500 years ago, may well end history and start something completely different. This book tells the story of how these three revolutions have affected humans and their fellow organisms.

There were humans long before there was history. Animals much like modern humans first appeared about 2.5 million years ago. But for countless generations they did not stand out from the myriad other organisms with which they shared their habitats.


txt = """The scientific objective of geodesy is to determine the size and shape of
the Earth. The practical role of geodesy is to provide a network of accurately
surveyed points on the Earth’s surface, the vertical elevations and geographic
positions of which are precisely known and, in turn, may be incorporated in
maps. When two geographic coordinates of a control point on the Earth’s
surface, its latitude and longitude, are known as well as its elevation above
sea level, the location of that point is known with accuracy within the limits
of error involved in the surveying processes. In mapping large areas, such as
a whole state or country, the irregularities in the curvature of the Earth must
be considered. A network of precisely surveyed control points provides a
skeleton to which other surveys may be tied to provide progressively finer
networks of more closely space points. The resulting networks of points have
many uses, including anchor points or bench marks for surveys of highways
and other civil features. A major use of control points is to provide reference
points to which the contour lines and other features of topographic maps are
tied. Most topographic maps are made using photogrammetric techniques
and aerial photographs."""